core/domain/html_cleaner.py

```diff
@@ -782,11 +783,202 @@ def add_caption_attr_to_image(html_string):
         str. Updated HTML string with the caption attribute for all
             oppia-noninteractive-image tags.
     """
-    soup = bs4.BeautifulSoup(html_string.encode('utf-8'), 'html.parser')
+    soup = bs4.BeautifulSoup(
+        html_string.encode(encoding='utf-8'), 'html.parser')
 
     for image in soup.findAll(name='oppia-noninteractive-image'):
         attrs = image.attrs
         if 'caption-with-value' not in attrs:
             image['caption-with-value'] = escape_html(json.dumps(''))
 
     return unicode(soup)
+
+
+def validate_customization_args(html_list):
+    """Validates customization arguments of Rich Text Components in a list of
+    html string.
+
+    Args:
+        html_list: list(str). List of html strings to be validated.
+
+    Returns:
+        dict: Dictionary of all the invalid customisation args where
+            key is a Rich Text Component and value is the invalid html string.
+    """
+    # Dictionary to hold html strings in which customization arguments
+    # are invalid.
+    err_dict = {}
+
+    for html_string in html_list:
+        soup = bs4.BeautifulSoup(
+            html_string.encode(encoding='utf-8'), 'html.parser')
+        _validate_customization_args_in_soup(soup, err_dict, html_string)
+
+        for collapsible in soup.findAll(
+                name='oppia-noninteractive-collapsible'):
+            # This is wrapped in try catch block since any invalid
+            # content such as int will not be convertible in Beautiful
+            # Soup format.
+            try:
+                content_html = json.loads(
+                    unescape_html(collapsible['content-with-value']))
+                soup_for_collapsible = bs4.BeautifulSoup(
+                    content_html, 'html.parser')
+                _validate_customization_args_in_soup(
```

Comment:
Does this need to be recursive, e.g. what happens if there's an image in collapsible in collapsible? If that's not allowed then there should be an additional check to ensure that the contents of collapsible/tabs don't themselves contain collapsibles/tabs, right?

Reply:
No, this should not be recursive because any nested collapsible or tab is invalid and checking any further content inside it is not required. It will be already added to the list of invalid strings.

Also, there is a check to find nested tabs and collapsible components here: https://github.com/oppia/oppia/pull/5362/commits/db188d558ac5831e13a5ef1a52955c5ad269a02f#diff-cfd52fa7bb3edfcd2ea1db12835eca82R862

Reply:
Also, the code written now is not recursive since any nested collapsible or tabs will be an attribute value i.e. they will not be present as escaped html to be considered as tags by BeautifulSoup.

Reply:
Hi @seanlip, I have a doubt here:

The first function converts a string to soup and passes the soup to second function for validation.

Further, if the soup has collapsible or tab components, it will convert the html content inside these components into a soup and again pass that for validation.

So, do you want the function to only prepare a list of those html strings which have some oppia-noninteractive-* tag and then pass that list to the second function?

Reply:
The first function should pass a list of the tags themselves, which can be represented as strings or as BeautifulSoup tags.

Reply:
So, in that case why do we need two functions, we can just have a single function which already has a list of html strings and calls itself recursively for content inside tabs and collapsible?

Reply:
The first function takes an arbitrary html string as an argument.

The second function takes a single oppia-noninteractive-* tag.

Reply:
The first function accepts a complete list of html strings from an exploration.

I store the invalid strings in a dict where keys are tags and the values are strings invalid for these tags.

So should I do this:

- First function accepts a list of html strings
- It then extracts various components from the strings and passes each component along with the string it belongs to the second function.
- The second function returns a bool whether the string is invalid for the passed component.
- The first function stores the invalid string in a dict along with the tag name for which it is invalid.

Reply:
The second function probably wants to return more info than a bool (perhaps some error string), otherwise it'll be harder to debug issues.

Doesn't really matter how the first function stores the strings internally, anything reasonable should probably work. Probably returns a dict of html tag strings mapped to error messages.

Reply:
I have made the changes, PTAL!